### ISSUE TITLE
agoo:  use of undeclared identifier 'SO_REUSEPORT'

### ISSUE THREAD

**Issue by deepj** [oracle/truffleruby#1592]
This issue is associated to #1566 and a support of pthread in sulong.

To reproduce:

```ruby
require 'agoo'
```

Error:

```
ERROR: pthread_create is unsupported!
truffleruby: org.graalvm.polyglot.PolyglotException: com.oracle.truffle.llvm.runtime.LLVMExitException
Original Internal Error:
com.oracle.truffle.llvm.runtime.LLVMExitException
```

**Comment by deepj** [comment]
@eregon I guess, this can be closed as well since sulong support pthread

**Comment by deepj** [comment]
I'm just getting a different issue

```
bind.c:289:35: error: use of undeclared identifier 'SO_REUSEPORT'
    setsockopt(b->fd, SOL_SOCKET, SO_REUSEPORT, &optval, sizeof(optval));
                                  ^
```

**Comment by eregon** [comment]
From `man 7 socket`:
```
SO_REUSEPORT (since Linux 3.9)
```
Which Linux version do you have?

And https://github.com/ohler55/agoo/blob/177154260c9323a789e78826a3a34bb2480aad76/ext/agoo/bind.c#L289 seems to reference `SO_REUSEPORT` unconditionally.

Does `agoo` install fine on your machine with MRI?

**Comment by eregon** [comment]
FWIW, we currently support Puma relatively well, but other webservers probably not as good.

**Comment by eregon** [comment]
It works for me locally:
```
$ gem i agoo 
Fetching agoo-2.11.5.gem
Building native extensions. This could take a while...
Successfully installed agoo-2.11.5
1 gem installed
```

So I think it's a bug of `agoo` when installed on older operating systems (or when installed on non-Linux).
Please report it there.

**Comment by deepj** [comment]
I'm on macOS. No issue under MRI. Compiled and worked. That's was why I reported it.

**Comment by eregon** [comment]
@pitr-ch Can you try to reproduce on macOS?

**Comment by pitr-ch** [comment]
I've reproduced it on macOS. I'll continue investigating when I am back from vacation.

**Comment by norswap** [comment]
Seems to be a consequence of the macro `OSX_OS` being undefined: https://github.com/ohler55/agoo/blob/develop/ext/agoo/bind.c#L286-L288

That macro is not defined in the agoo itself, and I found no reference to it in mkmf... or anywhere indexed by google, really.

I will try to find out why it works on MRI.

**Comment by norswap** [comment]
In Apple's `sys/socket.h`:

```
#if !defined(_POSIX_C_SOURCE) || defined(_DARWIN_C_SOURCE)
#define SO_REUSEPORT    0x0200          /* allow local address & port reuse */
```

Comparing the Makefile generated by MRI's mkmf and ours, they define `_DARWIN_C_SOURCE`, and we don't. I'll try to see what I can do to fix that.

In passing, this is probably a bug in agoo: `SO_REUSEPORT` has different semantics in recent Linux than in BSD & friends: https://stackoverflow.com/questions/41247790/

**Comment by norswap** [comment]
Fixed in 949cb6e7acd
